### PR TITLE
Dpe 39 memcached handle install event

### PR DESCRIPTION
**Pull request by quiet-ranger**
This PR is very large for two reasons:
1. It is the first one after automatically generating the charm structure
2. It includes some charm libraries

You only need to review two files:
- charm.py
- test_charm.py

For the latter, it would be valuable if you checked the following:
1. The class being tested is not mocked
2. All possible paths of the tested code are exercised, including exception ones
3. As far as possible, we avoid testing the implementation itself preferring to test the outcome instead

### CODE REVIEW

**Review by jnsgruk — COMMENTED** [review]
Mostly nits in comments, overall looking like a good start 😀

**Comment by jnsgruk** [comment]
Also I've just realised you didn't request this review and I just misunderstood GitHub nagging me about changes being pushed! Sorry if this is still in progress! 😅

File: src/charm.py [hunk 1]
```diff
@@ -12,76 +12,93 @@
...
 
-        Learn more about Pebble layers at https://github.com/canonical/pebble
+        Learn more about config at https://juju.is/docs/sdk/config
+        """
+        self.unit.status = MaintenanceStatus("Installing memcached")
+        try:
+            apt.add_package("memcached", update_cache=True, version="1.5.22-2ubuntu0.2")
+        except apt.PackageNotFoundError:
+            logger.error("memcached not found on repository")
+            self.unit.status = BlockedStatus("Failed to find memcached")
```
[the remark this answers is not in the document]
Comment (quiet-ranger) [reply src/charm.py:58]: Changed

File: requirements-dev.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+-r requirements.txt
```
[the remark this answers is not in the document]
Comment (quiet-ranger) [reply requirements-dev.txt:1]: Done

File: run_tests [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh -e
```
[the remark this answers is not in the document]
Comment (quiet-ranger) [reply run_tests:1]: Done

File: tox.ini [hunk 4]
```diff
@@ -47,21 +48,22 @@ commands =
...
 description = Run unit tests
 deps =
     pytest
-    coverage[toml]
-    -r{toxinidir}/requirements.txt
+    coverage[toml] 
+    -r{toxinidir}/requirements-dev.txt
 commands =
-    coverage run --source={[vars]src_path} \
+    coverage run --source={[vars]src_path},{[vars]lib_path} \
```
Comment (zmraul) [inline tox.ini:66]: I'm not a 100% percent sure on this, but I think the `lib_path` should be used if the charm implements a library. In this case the one you are using is an external one. I don't think we need to run coverage for the external libraries we use. Maybe @jnsgruk can correct me on this?

File: tox.ini [hunk 5]
```diff
@@ -9,8 +9,9 @@ envlist = lint, unit
 [vars]
 src_path = {toxinidir}/src/
 tst_path = {toxinidir}/tests/
-;lib_path = {toxinidir}/lib/charms/operator_name_with_underscores
-all_path = {[vars]src_path} {[vars]tst_path} 
+lib_path = {toxinidir}/lib/charms/operator_libs_linux
```
Comment (zmraul) [inline tox.ini:12]: You can comment out this one, as it is not used

File: src/charm.py [hunk 6]
```diff
@@ -12,76 +12,93 @@
...
 
-    def _on_httpbin_pebble_ready(self, event):
-        """Define and start a workload using the Pebble API.
+        self._stored.set_default(port="11211")
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        You'll need to specify the right entrypoint and environment
-        configuration for your specific workload. Tip: you can see the
-        standard entrypoint of an existing container using docker inspect
+    def _on_install(self, evt) -> None:
```
[the remark this answers is not in the document]
Comment (zeljkoagic) [reply src/charm.py:46]: This one has been a source of minor controversy: Exploiting `_` to denote unused variables is very heavily used and in my view very well received among practitioners, yet PEP 640 was [rejected by the Steering Council](https://www.python.org/dev/peps/pep-0640/). Not sure if there was another PEP afterwards, though. I am personally in favor of using `_` as it provides focus on what matters, and removes focus from that which is not relevant or ends up being unused further in the code base.

**Review by zeljkoagic — APPROVED** [review]

**Review by zeljkoagic — DISMISSED** [review]

**Review by jnsgruk — APPROVED** [review]

File: src/charm.py [hunk 7]
```diff
@@ -12,76 +12,91 @@
...
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
+        self.unit.status = MaintenanceStatus("Installing memcached")
+        try:
+            apt.add_package("memcached", update_cache=True, version="1.5.22-2ubuntu0.2")
+        except apt.PackageNotFoundError:
+            logger.error("memcached not found in repository")
+            self.unit.status = BlockedStatus("Could not install memcached")
+
+    def _on_start(self, event: StartEvent) -> None:
```
Comment (marceloneppel) [inline src/charm.py:58]: You can change `event` parameter to `_` (and remove the arg in the docstring) as it's not used in the code. However, ignore this suggestion if it's intended to be used in the next PRs. The same applies to L84.

File: lib/charms/operator_libs_linux/v0/systemd.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
+# Copyright 2021 Canonical Ltd.
```
Comment (paulomach) [inline lib/charms/operator_libs_linux/v0/systemd.py:1]: @quiet-ranger since this belongs to external lib, shouldn't it be excluded from this repo?

**Review by paulomach — COMMENTED** [review]
I'm just in doubt about the copy of files from upstream operators_libs. 
As @quiet-ranger , it's a matter of enabling CI tests to succeed. But IMHO, shouldn't it be an step in the test pipeline to acquire these libs? 
@jnsgruk can you shed a light?

File: requirements.txt [hunk 9]
```diff
@@ -1 +0,0 @@
-ops >= 1.2.0
```
Comment (marceloneppel) [inline requirements.txt:1]: I think we need this dependency when deploying the charm with `juju deploy`. Why was it removed?

**Review by paulomach — APPROVED** [review]

**Review by marceloneppel — APPROVED** [review]
LGTM Juan! I'm just wondering why ops dependency was removed and if it wouldn't lead to a missing dependency issue when deploying the charm.

**Comment by jnsgruk** [comment]
> I'm just in doubt about the copy of files from upstream operators_libs. As @quiet-ranger , it's a matter of enabling CI tests to succeed. But IMHO, shouldn't it be an step in the test pipeline to acquire these libs? @jnsgruk can you shed a light?

This is actually correct how it is now. The idea is to check in the library to your code as Juan has done here.

File: lib/charms/operator_libs_linux/v0/systemd.py (hunk 8 above)
Comment (quiet-ranger) [reply lib/charms/operator_libs_linux/v0/systemd.py:1]: Pending discussion with the rest of the team.  In the MongoDB-operator project, the charm library was also copied locally

**Comment by quiet-ranger** [comment]
All major issues resolved and plenty of minor comments taken into account.  There remains only some questions around "external" libraries

**Pull request merged**